pkg/chunk/chunk_store_utils.go

```diff
@@ -252,3 +256,10 @@ func (c *Fetcher) processCacheResponse(chunks []Chunk, keys []string, bufs [][]b
 	}
 	return found, missing, err
 }
+
+func injectShardLabels(chunks []Chunk, shard astmapper.ShardAnnotation) {
+	for i, chunk := range chunks {
+		chunk.Metric = append(chunk.Metric, shard.Label())
```

Comment:
Agree with Marco – better be safe and adhere to contract then break it. Suggested code looks good, although given that labels are supposedly sorted already, simple iteration to find the insertion point would do the trick as well. `__cortex_shard__` will most likely go first anyway (but may not).

Reply:
Thanks for the example code, I'll definitely incorporate it. I believe a reason why this has not come up is because these shard labels are later filtered out (we don't return `__cortex_shard__` labels in responses).